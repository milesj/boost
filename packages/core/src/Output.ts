/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import ansiEscapes from 'ansi-escapes';

export default class Output {
  protected completed: boolean = false;

  protected content: string = '';

  protected lastHeight: number = 0;

  constructor(content: string) {
    this.update(content);
  }

  isComplete(): boolean {
    return this.completed;
  }

  update(content: string) {
    this.content = content;
  }

  render(): string {
    const output = ansiEscapes.eraseLines(this.lastHeight) + this.content;

    this.lastHeight = this.content.split('\n').length;

    return output;
  }
}
