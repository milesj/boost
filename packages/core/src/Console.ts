/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import exit from 'exit';
import cliSize from 'term-size';
import ansiEscapes from 'ansi-escapes';
import Emitter from './Emitter';
import Tool from './Tool';
import { Debugger } from './types';
import Output from './Output';

// 16 FPS (60 FPS is actually too fast as it tears)
export const FPS_RATE = 62.5;

// Bind our writable streams for easy access
export const BOUND_WRITERS = {
  stderr: process.stderr.write.bind(process.stderr),
  stdout: process.stdout.write.bind(process.stdout),
};

export const WRAPPED_STREAMS = {
  stderr: false,
  stdout: false,
};

export type StreamType = 'stderr' | 'stdout';

export interface ConsoleState {
  disabled: boolean;
  final: boolean;
  started: boolean;
  stopped: boolean;
}

export default class Console extends Emitter {
  bufferedStreams: (() => void)[] = [];

  debug: Debugger;

  errorLogs: string[] = [];

  logs: string[] = [];

  outputQueue: Output[] = [];

  tool: Tool<any>;

  protected renderTimer: NodeJS.Timer | null = null;

  protected restoreCursorOnExit: boolean = false;

  protected state: ConsoleState = {
    disabled: false,
    final: false,
    started: false,
    stopped: false,
  };

  private writers: typeof BOUND_WRITERS;

  constructor(tool: Tool<any>, /* test only */ testWriters: typeof BOUND_WRITERS = BOUND_WRITERS) {
    super();

    this.debug = tool.createDebugger('console');
    this.tool = tool;
    this.writers = testWriters;

    // istanbul ignore next
    if (process.env.NODE_ENV !== 'test') {
      process
        .on('SIGINT', this.handleSignal)
        .on('SIGTERM', this.handleSignal)
        .on('uncaughtException', this.handleFailure)
        .on('unhandledRejection', this.handleFailure);
    }
  }

  /**
   * Disable the render loop entirely.
   */
  disable(): this {
    this.state.disabled = true;

    return this;
  }

  /**
   * Display a footer after all output.
   */
  displayFooter() {
    const { footer } = this.tool.options;

    if (footer) {
      this.out(footer, 1);
    }
  }

  /**
   * Display a header before all output.
   */
  displayHeader() {
    const { header } = this.tool.options;

    if (header) {
      this.out(header, 1);
    }
  }

  /**
   * Enable the render loop.
   */
  enable(): this {
    this.state.disabled = false;

    return this;
  }

  /**
   * Write a message to `stderr` with optional trailing newline(s).
   */
  err(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stderr(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Flush buffered stream output.
   */
  flushBufferedStreams(): this {
    this.bufferedStreams.forEach(buffer => {
      buffer();
    });

    return this;
  }

  /**
   * Flush the top output block in the queue.
   */
  flushOutputQueue(): this {
    const outputs = this.outputQueue.filter((out, i) => i === 0 || out.isConcurrent());

    // Erase the previous output
    outputs.forEach(output => {
      output.erasePrevious();
    });

    // Write buffered output
    this.flushBufferedStreams();

    // Write the next output
    outputs.forEach(output => {
      output.render();
    });

    // Remove completed outputs
    this.outputQueue = this.outputQueue.filter(out => !out.isComplete());

    // Stop the render loop once the queue is empty
    if (this.isEmptyQueue()) {
      this.stopRenderLoop();
    } else {
      this.startRenderLoop();
    }

    return this;
  }

  /**
   * Handle uncaught exceptions and unhandled rejections that bubble up.
   */
  handleFailure = (error: Error) => {
    this.start();
    this.debug('Uncaught exception or unresolved promise handled');
    this.stop(error);

    exit(2);
  };

  /**
   * Handle SIGINT and SIGTERM interruptions.
   */
  handleSignal = () => {
    this.start();
    this.debug('SIGINT or SIGTERM handled');
    this.stop(new Error(this.tool.msg('errors:processTerminated')));

    exit(2);
  };

  /**
   * Hide the console cursor.
   */
  hideCursor(): this {
    this.out(ansiEscapes.cursorHide);

    if (!this.restoreCursorOnExit && !this.isSilent()) {
      this.restoreCursorOnExit = true;

      // istanbul ignore next
      process.on('exit', () => {
        process.stdout.write(ansiEscapes.cursorShow);
      });
    }

    return this;
  }

  /**
   * Return true if the render loop has been disabled.
   */
  isDisabled(): boolean {
    return this.state.disabled;
  }

  /**
   * Return true if the output queue is empty.
   */
  isEmptyQueue(): boolean {
    return this.outputQueue.length === 0;
  }

  /**
   * Return true if the final render.
   */
  isFinalRender(): boolean {
    return this.state.final;
  }

  /**
   * Return true if the there should be no output.
   */
  isSilent(): boolean {
    return this.tool.config.silent;
  }

  /**
   * Return true if the defined stream has been wrapped by the console layer.
   */
  isStreamWrapped(type: StreamType): boolean {
    return !!WRAPPED_STREAMS[type];
  }

  /**
   * Log a message to display on success during the final render.
   */
  log(message: string): this {
    this.logs.push(message);

    return this;
  }

  /**
   * Log a live message to display during the rendering process.
   */
  logLive(message: string): this {
    // Write to the wrapped buffer
    process.stdout.write(message);

    return this;
  }

  /**
   * Log an error message to display on failure during the final render.
   */
  logError(message: string): this {
    this.errorLogs.push(message);

    return this;
  }

  /**
   * Write a message to `stdout` with optional trailing newline(s).
   */
  out(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stdout(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Enqueue a block of output to be rendered.
   */
  render(output: Output): this {
    if (this.isDisabled()) {
      throw new Error(
        'Output cannot be enqueued as the render loop has been disabled. This is usually caused by conflicting reporters.',
      );
    }

    if (!this.outputQueue.includes(output)) {
      this.outputQueue.push(output);
    }

    // Only run the render loop when output is enqueued
    this.startRenderLoop();

    return this;
  }

  /**
   * Handle the final rendering of all output before stopping.
   */
  renderFinalOutput(error: Error | null) {
    this.debug('Rendering final console output');
    this.state.final = true;

    // Mark all output as final
    this.outputQueue.forEach(output => {
      output.enqueue(true);
    });

    // Recursively render the remaining output
    while (!this.isEmptyQueue()) {
      this.flushOutputQueue();
    }

    // Stop the render loop
    this.stopRenderLoop();

    if (error) {
      if (this.errorLogs.length > 0) {
        this.err(`\n${this.errorLogs.join('\n')}\n`);
      }

      this.emit('error', [error]);
    } else {
      if (this.logs.length > 0) {
        this.out(`\n${this.logs.join('\n')}\n`);
      }

      this.displayFooter();
    }

    // Flush any stream output that still exists
    this.flushBufferedStreams();

    // Show the cursor incase it has been hidden
    this.out(ansiEscapes.cursorShow);
  }

  /**
   * Reset the cursor back to the bottom of the console.
   */
  resetCursor(): this {
    this.out(ansiEscapes.cursorTo(0, cliSize().rows));

    return this;
  }

  /**
   * Show the console cursor.
   */
  showCursor(): this {
    this.restoreCursorOnExit = false;
    this.out(ansiEscapes.cursorShow);

    return this;
  }

  /**
   * Start the console by wrapping streams and buffering output.
   */
  start(args: any[] = []): this {
    if (this.state.started) {
      return this;
    }

    this.debug('Starting console render loop');
    this.emit('start', args);
    this.wrapStreams();
    this.displayHeader();
    this.state.started = true;

    return this;
  }

  /**
   * Automatically render the console in a timeout loop at 16 FPS.
   */
  startRenderLoop() {
    if (this.isSilent() || this.isDisabled()) {
      return;
    }

    this.renderTimer = setTimeout(() => {
      this.flushOutputQueue();
    }, FPS_RATE);
  }

  /**
   * Stop the console rendering process.
   */
  stop(error: Error | null = null) {
    if (this.state.stopped) {
      return;
    }

    if (error) {
      this.debug('Stopping console with an error');
      this.errorLogs.push(...this.logs);
      this.logs = [];
    } else {
      this.debug('Stopping console render loop');
    }

    this.renderFinalOutput(error);
    this.unwrapStreams();
    this.emit('stop', [error]);
    this.state.stopped = true;
    this.state.started = false;
  }

  /**
   * Stop the background render loop.
   */
  stopRenderLoop() {
    if (this.renderTimer) {
      clearTimeout(this.renderTimer);
      this.renderTimer = null;
    }
  }

  /**
   * Unwrap the native console and reset it back to normal.
   */
  unwrapStreams() {
    if (this.isSilent() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;

      if (!this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Unwrapping `%s` stream', name);

      process[name].write = this.writers[name] as any;

      WRAPPED_STREAMS[name] = false;
    });
  }

  /**
   * Wrap the `stdout` and `stderr` streams and buffer the output as
   * to not collide with our render loop.
   */
  wrapStreams() {
    if (this.isSilent() || this.isDisabled() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;
      const stream = process[name];
      let buffer = '';

      if (this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Wrapping `%s` stream', name);

      this.bufferedStreams.push(() => {
        if (stream.isTTY && buffer) {
          this.writers[name](buffer);
        }

        buffer = '';
      });

      stream.write = (chunk: string) => {
        // No output, display immediately
        if (this.isEmptyQueue()) {
          this.writers[name](chunk);
        } else {
          buffer += String(chunk);
        }

        return true;
      };

      WRAPPED_STREAMS[name] = true;
    });
  }
}
