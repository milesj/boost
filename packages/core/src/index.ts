/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import CLI from './CLI';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Context from './Context';
import Emitter, { EventArguments, EventListener } from './Emitter';
import ModuleLoader from './ModuleLoader';
import Pipeline from './Pipeline';
import Plugin from './Plugin';
import Reporter from './Reporter';
import Routine from './Routine';
import Task from './Task';
import Tool, { ToolOptions, ToolConfig, ToolPluginRegistry } from './Tool';

export {
  CLI,
  ConfigLoader,
  Console,
  Context,
  Emitter,
  EventArguments,
  EventListener,
  ModuleLoader,
  Pipeline,
  Plugin,
  Reporter,
  Routine,
  Task,
  Tool,
  ToolConfig,
  ToolOptions,
  ToolPluginRegistry,
};

export * from './types';
