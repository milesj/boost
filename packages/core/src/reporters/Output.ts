/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import ansiEscapes from 'ansi-escapes';
import Console from '../Console';

export default class Output {
  console: Console;

  content: string = '';

  lastHeight: number = 0;

  constructor(console: Console) {
    this.console = console;
  }

  /**
   * Update the output block with the defined content and render the console.
   * Since an output represents an exact line, or a collection of lines,
   * we must always end with a new line to calculate height correctly.
   */
  update(content: string) {
    this.content = content;

    // Always end with a new line
    if (!this.content.endsWith('\n')) {
      this.content += '\n';
    }

    this.console.render();
  }

  /**
   * Render the content to the console. If this output has previously
   * been rendered, re-render it by erasing previous output.
   */
  render(): string {
    let output = ansiEscapes.eraseLines(this.lastHeight);

    // Count the height of the content without the ANSI escape codes above
    this.lastHeight = this.content.split('\n').length;

    output += this.content;

    return output;
  }
}
