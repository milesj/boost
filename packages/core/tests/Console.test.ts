import exit from 'exit';
import { mockTool } from '../src/tests';
import Console, { WRAPPED_STREAMS } from '../src/Console';
import Output from '../src/Output';
import { SignalError } from '../src';

jest.mock('exit');

describe('Console', () => {
  let cli: Console;
  let err: jest.Mock;
  let out: jest.Mock;

  beforeEach(() => {
    err = jest.fn();
    out = jest.fn();
    cli = new Console(mockTool(), {
      stderr: err,
      stdout: out,
    });
  });

  describe('disable()', () => {
    it('marks as disabled', () => {
      expect(cli.isDisabled()).toBe(false);

      cli.disable();

      expect(cli.isDisabled()).toBe(true);
    });
  });

  describe('displayFooter()', () => {
    it('displays nothing if no footer', () => {
      cli.displayFooter();

      expect(out).not.toHaveBeenCalled();
    });

    it('displays the footer', () => {
      cli.tool.options.footer = 'Powered by Boost';
      cli.displayFooter();

      expect(out).toHaveBeenCalledWith('Powered by Boost\n');
    });
  });

  describe('displayHeader()', () => {
    it('displays nothing if no header', () => {
      cli.displayHeader();

      expect(out).not.toHaveBeenCalled();
    });

    it('displays the header', () => {
      cli.tool.options.header = 'Powered by Boost';
      cli.displayHeader();

      expect(out).toHaveBeenCalledWith('Powered by Boost\n');
    });
  });

  describe('enable()', () => {
    it('marks as enabled', () => {
      cli.disable();

      expect(cli.isDisabled()).toBe(true);

      cli.enable();

      expect(cli.isDisabled()).toBe(false);
    });
  });

  describe('err()', () => {
    it('writes a message to `stderr`', () => {
      cli.err('Hello');

      expect(err).toHaveBeenCalledWith('Hello');
    });

    it('doesnt write to `stderr` if config is silent', () => {
      cli.tool.config.silent = true;
      cli.err('Hello');

      expect(err).not.toHaveBeenCalled();
    });

    it('can customize the number of trailing newlines', () => {
      cli.err('Hello', 3);
      cli.err('Hello', 1);

      expect(err).toHaveBeenCalledWith('Hello\n\n\n');
      expect(err).toHaveBeenCalledWith('Hello\n');
    });
  });

  describe('flushBufferedStreams()', () => {
    it('calls all buffer callbacks', () => {
      const spy1 = jest.fn();
      const spy2 = jest.fn();

      cli.bufferedStreams.push(spy1, spy2);
      cli.flushBufferedStreams();

      expect(spy1).toHaveBeenCalled();
      expect(spy2).toHaveBeenCalled();
    });
  });

  describe('flushOutputQueue()', () => {
    it('flushes buffered streams', () => {
      const spy = jest.spyOn(cli, 'flushBufferedStreams');

      cli.flushOutputQueue();

      expect(spy).toHaveBeenCalled();
    });

    it('re-flushes the queue when output is complete and queue isnt empty', () => {
      const loopSpy = jest.spyOn(cli, 'startRenderLoop');
      const o1 = new Output(cli, () => 'foo');
      const o2 = new Output(cli, () => 'bar');

      o1.enqueue(true);

      cli.outputQueue.push(o1, o2);
      cli.flushOutputQueue();

      expect(loopSpy).toHaveBeenCalledTimes(2);
      expect(out).toHaveBeenCalledWith('foo\n');
      expect(cli.outputQueue).toEqual([o2]);
    });

    it('stops the loop when output is complete and queue is empty', () => {
      const loopSpy = jest.spyOn(cli, 'stopRenderLoop');
      const output = new Output(cli, () => 'foo');

      output.enqueue(true);

      cli.outputQueue.push(output);
      cli.flushOutputQueue();

      expect(loopSpy).toHaveBeenCalledTimes(1);
      expect(out).toHaveBeenCalledWith('foo\n');
      expect(cli.outputQueue).toEqual([]);
    });

    describe('non-concurrent', () => {
      it('renders an output', () => {
        const output = new Output(cli, () => 'foo');
        const eraseSpy = jest.spyOn(output, 'erasePrevious');
        const renderSpy = jest.spyOn(output, 'render');

        cli.outputQueue.push(output);
        cli.flushOutputQueue();

        expect(eraseSpy).toHaveBeenCalled();
        expect(renderSpy).toHaveBeenCalled();
        expect(out).toHaveBeenCalledWith('foo\n');
        expect(cli.outputQueue).toEqual([output]);
      });

      it('only renders the first output', () => {
        const o1 = new Output(cli, () => 'foo');
        const o2 = new Output(cli, () => 'bar');
        const eraseSpy = jest.spyOn(o2, 'erasePrevious');
        const renderSpy = jest.spyOn(o2, 'render');

        cli.outputQueue.push(o1, o2);
        cli.flushOutputQueue();

        expect(eraseSpy).not.toHaveBeenCalled();
        expect(renderSpy).not.toHaveBeenCalled();
        expect(out).toHaveBeenCalledWith('foo\n');
        expect(cli.outputQueue).toEqual([o1, o2]);
      });

      it('removes the output when complete', () => {
        const output = new Output(cli, () => 'foo');
        output.enqueue(true);

        cli.outputQueue.push(output);
        cli.flushOutputQueue();

        expect(out).toHaveBeenCalledWith('foo\n');
        expect(cli.outputQueue).toEqual([]);
      });
    });

    describe('concurrent', () => {
      it('renders first output and concurrent outputs', () => {
        const o1 = new Output(cli, () => 'foo');
        const o2 = new Output(cli, () => 'bar');
        const o3 = new Output(cli, () => 'baz').concurrent();

        cli.outputQueue.push(o1, o2, o3);
        cli.flushOutputQueue();

        expect(out).toHaveBeenCalledWith('foo\n');
        expect(out).toHaveBeenCalledWith('baz\n');
        expect(cli.outputQueue).toEqual([o1, o2, o3]);
      });

      it('renders all concurrent outputs', () => {
        const o1 = new Output(cli, () => 'foo');
        const o2 = new Output(cli, () => 'bar');
        const o3 = new Output(cli, () => 'baz').concurrent();
        const o4 = new Output(cli, () => 'qux');
        const o5 = new Output(cli, () => 'wtf').concurrent();

        cli.outputQueue.push(o1, o2, o3, o4, o5);
        cli.flushOutputQueue();

        expect(out).toHaveBeenCalledWith('foo\n');
        expect(out).toHaveBeenCalledWith('baz\n');
        expect(out).toHaveBeenCalledWith('wtf\n');
        expect(cli.outputQueue).toEqual([o1, o2, o3, o4, o5]);
      });

      it('removes all concurrent outputs that are completed', () => {
        const o1 = new Output(cli, () => 'foo');
        const o2 = new Output(cli, () => 'bar');
        const o3 = new Output(cli, () => 'baz').concurrent();
        const o4 = new Output(cli, () => 'qux');
        const o5 = new Output(cli, () => 'wtf').concurrent();

        o1.enqueue(true);
        o5.enqueue(true);

        cli.outputQueue.push(o1, o2, o3, o4, o5);
        cli.flushOutputQueue();

        expect(out).toHaveBeenCalledWith('foo\n');
        expect(out).toHaveBeenCalledWith('baz\n');
        expect(out).toHaveBeenCalledWith('wtf\n');
        expect(cli.outputQueue).toEqual([o2, o3, o4]);
      });
    });
  });

  describe('handleFailure()', () => {
    it('calls exit with error', () => {
      const error = new Error('Oops');

      cli.stop = jest.fn();
      cli.handleFailure(error);

      expect(cli.stop).toHaveBeenCalledWith(error);
    });

    it('exits with a code of 2', () => {
      cli.handleFailure(new Error('Oops'));

      expect(exit).toHaveBeenCalledWith(2);
    });
  });

  describe('handleSignal()', () => {
    it('calls exit with error', () => {
      cli.stop = jest.fn();
      cli.handleSignal('SIGINT');

      expect(cli.stop).toHaveBeenCalledWith(
        new SignalError('Process has been terminated.', 'SIGINT'),
      );
    });

    it('exits with a code of 2', () => {
      cli.handleSignal('SIGTERM');

      expect(exit).toHaveBeenCalledWith(2);
    });
  });

  describe('isSilent()', () => {
    it('returns value from config', () => {
      expect(cli.isSilent()).toBe(false);

      cli.tool.config.silent = true;

      expect(cli.isSilent()).toBe(true);
    });
  });

  describe('log()', () => {
    it('adds a log', () => {
      expect(cli.logs).toEqual([]);

      cli.log('foo');

      expect(cli.logs).toEqual(['foo']);
    });
  });

  describe('logLive()', () => {
    it('adds a log via `console.log`', () => {
      const old = process.stdout.write.bind(process.stdout);
      const spy = jest.fn();

      process.stdout.write = spy;

      cli.logLive('foo');

      expect(spy).toHaveBeenCalledWith('foo');

      // @ts-ignore
      process.stdout.write = old;
    });
  });

  describe('logError()', () => {
    it('adds a log', () => {
      expect(cli.errorLogs).toEqual([]);

      cli.logError('foo');

      expect(cli.errorLogs).toEqual(['foo']);
    });
  });

  describe('out()', () => {
    it('writes a message to `stdout`', () => {
      cli.out('Hello');

      expect(out).toHaveBeenCalledWith('Hello');
    });

    it('doesnt write to `stdout` if config is silent', () => {
      cli.tool.config.silent = true;
      cli.out('Hello');

      expect(out).not.toHaveBeenCalled();
    });

    it('can customize the number of trailing newlines', () => {
      cli.out('Hello', 3);
      cli.out('Hello', 1);

      expect(out).toHaveBeenCalledWith('Hello\n\n\n');
      expect(out).toHaveBeenCalledWith('Hello\n');
    });
  });

  describe('render()', () => {
    it('enqueues an `Output` and doesnt duplicate', () => {
      expect(cli.outputQueue).toEqual([]);

      const output = new Output(cli, () => 'foo');

      cli.render(output);

      expect(cli.outputQueue).toEqual([output]);

      cli.render(output);

      expect(cli.outputQueue).toEqual([output]);
    });

    it('throws an error when render loop is disabled', () => {
      cli.disable();

      expect(() => {
        cli.render(new Output(cli, () => 'foo'));
      }).toThrowErrorMatchingSnapshot();
    });

    it('starts the render loop', () => {
      const spy = jest.spyOn(cli, 'startRenderLoop');

      cli.render(new Output(cli, () => 'foo'));

      expect(spy).toHaveBeenCalled();
    });
  });

  describe('renderFinalOutput()', () => {
    it('flushes and stops loop', () => {
      const queueSpy = jest.spyOn(cli, 'flushOutputQueue');
      const streamSpy = jest.spyOn(cli, 'flushBufferedStreams');
      const loopSpy = jest.spyOn(cli, 'stopRenderLoop');

      cli.renderFinalOutput(null);

      expect(queueSpy).not.toHaveBeenCalled();
      expect(streamSpy).toHaveBeenCalled();
      expect(loopSpy).toHaveBeenCalled();
    });

    it('marks all output as final and renders', () => {
      const queueSpy = jest.spyOn(cli, 'flushOutputQueue');
      const foo = new Output(cli, () => 'foo');
      const bar = new Output(cli, () => 'bar');
      const baz = new Output(cli, () => 'baz');

      cli.outputQueue.push(foo, bar, baz);
      cli.renderFinalOutput(null);

      expect(queueSpy).toHaveBeenCalled();

      expect(foo.isFinal()).toBe(true);
      expect(bar.isFinal()).toBe(true);
      expect(baz.isFinal()).toBe(true);

      expect(foo.isComplete()).toBe(true);
      expect(bar.isComplete()).toBe(true);
      expect(baz.isComplete()).toBe(true);

      expect(out).toHaveBeenCalledWith('foo\n');
      expect(out).toHaveBeenCalledWith('bar\n');
      expect(out).toHaveBeenCalledWith('baz\n');
    });

    it('displays logs on success', () => {
      cli.log('foo');
      cli.log('bar');
      cli.log('baz');

      cli.renderFinalOutput(null);

      expect(out).toHaveBeenCalledWith('\nfoo\nbar\nbaz\n');
    });

    it('doesnt display logs on failure', () => {
      cli.log('foo');
      cli.log('bar');
      cli.log('baz');

      cli.renderFinalOutput(new Error());

      expect(out).not.toHaveBeenCalledWith('\nfoo\nbar\nbaz\n');
    });

    it('displays error logs on failure', () => {
      cli.logError('foo');
      cli.logError('bar');
      cli.logError('baz');

      cli.renderFinalOutput(new Error());

      expect(err).toHaveBeenCalledWith('\nfoo\nbar\nbaz\n');
    });

    it('doesnt display error logs on success', () => {
      cli.logError('foo');
      cli.logError('bar');
      cli.logError('baz');

      cli.renderFinalOutput(null);

      expect(err).not.toHaveBeenCalled();
    });

    it('displays footer on success', () => {
      cli.tool.options.footer = 'Footer';

      cli.renderFinalOutput(null);

      expect(out).toHaveBeenCalledWith('Footer\n');
    });

    it('doesnt display footer on failure', () => {
      cli.tool.options.footer = 'Footer';

      cli.renderFinalOutput(new Error());

      expect(out).not.toHaveBeenCalledWith('Footer');
    });

    it('emits `error` event on failure', () => {
      const spy = jest.fn();
      const error = new Error('Stop');

      cli.on('error', spy);
      cli.renderFinalOutput(error);

      expect(spy).toHaveBeenCalledWith(error);
    });

    it('doesnt emit `error` event on success', () => {
      const spy = jest.fn();

      cli.on('error', spy);
      cli.renderFinalOutput(null);

      expect(spy).not.toHaveBeenCalled();
    });

    it('sets `final` flag', () => {
      expect(cli.isFinalRender()).toBe(false);

      cli.renderFinalOutput(null);

      expect(cli.isFinalRender()).toBe(true);
    });
  });

  describe('start()', () => {
    it('sets `started` flag', () => {
      // @ts-ignore Allow access
      expect(cli.state.started).toBe(false);

      cli.start();

      // @ts-ignore Allow access
      expect(cli.state.started).toBe(true);
    });

    it('only triggers once if started', () => {
      const spy = jest.spyOn(cli, 'wrapStreams');

      cli.start();
      cli.start();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('triggers the initial process methods', () => {
      cli.wrapStreams = jest.fn();
      cli.displayHeader = jest.fn();
      cli.start([]);

      expect(cli.wrapStreams).toHaveBeenCalled();
      expect(cli.displayHeader).toHaveBeenCalled();
    });

    it('emits `start` event', () => {
      const spy = jest.fn();

      cli.on('start', spy);
      cli.wrapStreams = jest.fn();
      cli.displayHeader = jest.fn();
      cli.startRenderLoop = jest.fn();
      cli.start([1, 2, 3]);

      expect(spy).toHaveBeenCalledWith(1, 2, 3);
    });
  });

  describe('startRenderLoop()', () => {
    it('sets a timer', () => {
      // @ts-ignore Allow access
      expect(cli.renderTimer).toBeNull();

      cli.startRenderLoop();

      // @ts-ignore Allow access
      expect(cli.renderTimer).not.toBeNull();
    });

    it('doesnt set a timer if disabled', () => {
      cli.disable();
      cli.startRenderLoop();

      // @ts-ignore Allow access
      expect(cli.renderTimer).toBeNull();
    });

    it('doesnt set a timer if silent', () => {
      cli.tool.config.silent = true;
      cli.startRenderLoop();

      // @ts-ignore Allow access
      expect(cli.renderTimer).toBeNull();
    });

    it('flushes output and resets timer', () => {
      const spy = jest.spyOn(cli, 'flushOutputQueue');

      cli.startRenderLoop();

      // @ts-ignore Allow access
      expect(cli.renderTimer).not.toBeNull();

      jest.advanceTimersByTime(200);

      expect(spy).toHaveBeenCalledWith();

      // @ts-ignore Allow access
      expect(cli.renderTimer).toBeNull();
    });
  });

  describe('stop()', () => {
    it('sets `stopped` flag', () => {
      // @ts-ignore Allow access
      expect(cli.state.stopped).toBe(false);

      cli.stop();

      // @ts-ignore Allow access
      expect(cli.state.stopped).toBe(true);
    });

    it('sets `started` flag', () => {
      cli.start();

      // @ts-ignore Allow access
      expect(cli.state.started).toBe(true);

      cli.stop();

      // @ts-ignore Allow access
      expect(cli.state.started).toBe(false);
    });

    it('only triggers once if stopping', () => {
      const spy = jest.spyOn(cli, 'renderFinalOutput');

      cli.stop();
      cli.stop();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('calls `stop` with null (no error)', () => {
      const spy = jest.fn();

      cli.on('stop', spy);
      cli.stop();

      expect(spy).toHaveBeenCalledWith(null);
    });

    it('calls `stop` with error', () => {
      const spy = jest.fn();
      const error = new Error('Oops');

      cli.on('stop', spy);
      cli.stop(error);

      expect(spy).toHaveBeenCalledWith(error);
    });

    it('renders final output', () => {
      const spy = jest.spyOn(cli, 'renderFinalOutput');

      cli.stop();

      expect(spy).toHaveBeenCalledWith(null);
    });

    it('renders final output with an error', () => {
      const spy = jest.spyOn(cli, 'renderFinalOutput');

      cli.stop(new Error('Oops'));

      expect(spy).toHaveBeenCalledWith(new Error('Oops'));
    });

    it('unwraps streams', () => {
      const spy = jest.spyOn(cli, 'unwrapStreams');

      cli.stop();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('copies logs to error logs if persist is true', () => {
      cli.logs.push('Hello');

      expect(cli.errorLogs).toEqual([]);

      cli.stop(new Error('Oops'));

      expect(cli.logs).toEqual([]);
      expect(cli.errorLogs).toEqual(['Hello']);
    });

    it('doesnt copy logs if not an error', () => {
      cli.logs.push('Hello');

      expect(cli.errorLogs).toEqual([]);

      cli.stop();

      expect(cli.logs).toEqual(['Hello']);
      expect(cli.errorLogs).toEqual([]);
    });
  });

  describe('stopRenderLoop()', () => {
    it('it clears the timer', () => {
      // @ts-ignore
      cli.renderTimer = 123;
      cli.stopRenderLoop();

      // @ts-ignore Allow access
      expect(cli.renderTimer).toBeNull();
    });
  });

  describe('streams', () => {
    const oldErr = process.stderr.write;
    const oldOut = process.stdout.write;

    beforeEach(() => {
      process.env.NODE_ENV = 'test-streams';
    });

    afterEach(() => {
      process.env.NODE_ENV = 'test';

      process.stderr.write = oldErr;
      process.stdout.write = oldOut;

      WRAPPED_STREAMS.stderr = false;
      WRAPPED_STREAMS.stdout = false;
    });

    describe('unwrapStreams()', () => {
      it('sets process streams to defined writes', () => {
        cli.wrapStreams();
        cli.unwrapStreams();

        expect(process.stderr.write).toBe(err);
        expect(process.stdout.write).toBe(out);

        expect(cli.isStreamWrapped('stderr')).toBe(false);
        expect(cli.isStreamWrapped('stdout')).toBe(false);
      });

      it('doesnt set process streams to defined writes when silent', () => {
        cli.tool.config.silent = true;
        cli.wrapStreams();
        cli.unwrapStreams();

        expect(process.stderr.write).not.toBe(err);
        expect(process.stdout.write).not.toBe(out);
      });
    });

    describe('wrapStreams()', () => {
      it('sets process streams to custom writers', () => {
        cli.wrapStreams();

        expect(process.stderr.write).not.toBe(oldErr);
        expect(process.stdout.write).not.toBe(oldOut);

        expect(cli.isStreamWrapped('stderr')).toBe(true);
        expect(cli.isStreamWrapped('stdout')).toBe(true);
      });

      it('only wraps once per stream', () => {
        cli.wrapStreams();
        cli.wrapStreams();
        cli.wrapStreams();

        expect(cli.bufferedStreams).toHaveLength(2);
      });

      it('doesnt set process streams to custom writers when silent', () => {
        cli.tool.config.silent = true;
        cli.wrapStreams();

        expect(cli.bufferedStreams).toHaveLength(0);
      });

      it('doesnt set process streams to custom writers when disabled', () => {
        cli.disable();
        cli.wrapStreams();

        expect(cli.bufferedStreams).toHaveLength(0);
      });

      it('calls stream directly when queue is empty', () => {
        cli.wrapStreams();

        process.stdout.write('foo');

        expect(out).toHaveBeenCalledWith('foo');
      });

      it('buffers stream when queue is not empty', () => {
        cli.wrapStreams();
        cli.outputQueue.push(new Output(cli, () => 'test'));

        process.stdout.write('foo');

        expect(out).not.toHaveBeenCalledWith('foo');
      });
    });
  });
});
