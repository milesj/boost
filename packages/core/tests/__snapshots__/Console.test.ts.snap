// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console render() throws an error when render loop is disabled 1`] = `"Output cannot be enqueued as the render loop has been disabled. This is usually caused by conflicting reporters."`;
